controllers/operator/storage_controller.go

```diff
@@ -0,0 +1,274 @@
+		return ctrl.Result{}, err
+	}
+	if err := r.checkState(ctx, log, &storage); err != nil {
+		log.Error(err, "failed to check sub resources state")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{RequeueAfter: schedDuration}, nil
+}
+
+func (r *StorageReconciler) checkState(ctx context.Context, log logr.Logger, storage *operatorv1alpha1.Storage) error {
+	overlay := operatorv1alpha1.StorageStatus{}
+	statefulset := apps.StatefulSet{}
+	errCol := new(kubernetes.ErrorCollector)
+	object := client.ObjectKey{Namespace: storage.Namespace, Name: storage.Name + "-" + storage.Spec.Type}
+	if err := r.Client.Get(ctx, object, &statefulset); err != nil && !apierrors.IsNotFound(err) {
+		errCol.Collect(fmt.Errorf("failed to get statefulset: %w", err))
+	} else {
+		overlay.Conditions = statefulset.Status.Conditions
+	}
+
+	if apiequal.Semantic.DeepDerivative(overlay, storage.Status) {
+		log.Info("Status keeps the same as before")
+	}
+	storage.Status = overlay
+	storage.Kind = "Storage"
+	if err := kubernetes.ApplyOverlay(storage, &operatorv1alpha1.Storage{Status: overlay}); err != nil {
+		errCol.Collect(fmt.Errorf("failed to apply overlay: %w", err))
+		return errCol.Error()
+	}
+	if err := r.Status().Update(ctx, storage); err != nil {
+		errCol.Collect(fmt.Errorf("failed to update status of es: %w", err))
+	}
+	log.Info("updated Status sub resource")
+	return errCol.Error()
+}
+
+func (r *StorageReconciler) checkSecurity(ctx context.Context, log logr.Logger, s *operatorv1alpha1.Storage) {
+	user, tls := s.Spec.Security.User, s.Spec.Security.TLS
+	if user.SecretName != "" {
+		if user.SecretName == "default" {
+			s.Spec.Config = append(s.Spec.Config, core.EnvVar{Name: "ELASTIC_USER", Value: "elastic"})
+			s.Spec.Config = append(s.Spec.Config, core.EnvVar{Name: "ELASTIC_PASSWORD", Value: "changeme"})
+		} else {
+			usersecret := core.Secret{}
+			if err := r.Client.Get(ctx, client.ObjectKey{Namespace: s.Namespace, Name: user.SecretName}, &usersecret); err != nil && !apierrors.IsNotFound(err) {
+				log.Info("fail get usersecret ")
+			}
+			for k, v := range usersecret.Data {
+				if k == "username" {
+					s.Spec.Config = append(s.Spec.Config, core.EnvVar{Name: "ELASTIC_USER", Value: string(v)})
+				} else if k == "password" {
+					s.Spec.Config = append(s.Spec.Config, core.EnvVar{Name: "ELASTIC_PASSWORD", Value: string(v)})
+				}
+			}
+		}
+	}
+	if tls {
+		s.Spec.ServiceName = "skywalking-storage"
+	} else {
+		s.Spec.ServiceName = s.Name + "-" + s.Spec.Type
+	}
+	s.Spec.Config = append(s.Spec.Config, core.EnvVar{Name: "discovery.seed_hosts", Value: s.Spec.ServiceName})
+}
+
+func (r *StorageReconciler) createCert(ctx context.Context, log logr.Logger, s *operatorv1alpha1.Storage) {
+	clientset, err := kubeclient.NewForConfig(r.RestConfig)
+	if err != nil {
+		return
+	}
+	secret2, _ := clientset.CoreV1().Secrets(s.Namespace).Get(ctx, "skywalking-storage", metav1.GetOptions{})
+	if secret2.Name != "" {
+		return
```

Comment:
How to renew the cert when it's expired? 

Here we can warn the cert will expire soon in operator's console